cuda/solver/lower_trs_kernels.cu

```diff
@@ -53,10 +58,124 @@ namespace cuda {
+        cusparseDestroySolveAnalysisInfo(cusp_csrsm_data.solve_info));
+#endif
+}
+
+
 template <typename ValueType, typename IndexType>
 void generate(std::shared_ptr<const CudaExecutor> exec,
               const matrix::Csr<ValueType, IndexType> *matrix,
-              const matrix::Dense<ValueType> *b) GKO_NOT_IMPLEMENTED;
+              const gko::size_type num_rhs)
+{
+    if (cusparse::is_supported<ValueType, IndexType>::value) {
+        auto handle = exec->get_cusparse_handle();
+#if (defined(CUDA_VERSION) && (CUDA_VERSION > 9010))
+        ValueType one = 1.0;
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseCreateCsrsm2Info(&cusp_csrsm2_data.solve_info));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseCreateMatDescr(&cusp_csrsm2_data.factor_descr));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatIndexBase(
+            cusp_csrsm2_data.factor_descr, CUSPARSE_INDEX_BASE_ZERO));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatType(
+            cusp_csrsm2_data.factor_descr, CUSPARSE_MATRIX_TYPE_GENERAL));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatDiagType(
+            cusp_csrsm2_data.factor_descr, CUSPARSE_DIAG_TYPE_NON_UNIT));
+        cusp_csrsm2_data.algorithm = 0;
+        cusp_csrsm2_data.policy = CUSPARSE_SOLVE_POLICY_USE_LEVEL;
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseCreateSolveAnalysisInfo(&cusp_csrsm_data.solve_info));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseCreateMatDescr(&cusp_csrsm_data.factor_descr));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatIndexBase(
+            cusp_csrsm_data.factor_descr, CUSPARSE_INDEX_BASE_ZERO));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatType(
+            cusp_csrsm_data.factor_descr, CUSPARSE_MATRIX_TYPE_GENERAL));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseSetMatDiagType(
+            cusp_csrsm_data.factor_descr, CUSPARSE_DIAG_TYPE_NON_UNIT));
+#endif
+
+#if (defined(CUDA_VERSION) && (CUDA_VERSION > 9010))
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseSetPointerMode(handle, CUSPARSE_POINTER_MODE_HOST));
+        cusparse::buffer_size_ext(
+            handle, cusp_csrsm2_data.algorithm,
+            CUSPARSE_OPERATION_NON_TRANSPOSE, CUSPARSE_OPERATION_TRANSPOSE,
+            matrix->get_size()[0], num_rhs, matrix->get_num_stored_elements(),
+            &one, cusp_csrsm2_data.factor_descr, matrix->get_const_values(),
+            matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(), nullptr,
+            num_rhs, cusp_csrsm2_data.solve_info, cusp_csrsm2_data.policy,
+            &cusp_csrsm2_data.factor_work_size);
+
+        // allocate workspace
+        if (cusp_csrsm2_data.factor_work_vec != nullptr) {
+            exec->free(cusp_csrsm2_data.factor_work_vec);
+        }
+        cusp_csrsm2_data.factor_work_vec =
+            exec->alloc<void *>(cusp_csrsm2_data.factor_work_size);
```

Comment:
Please move these allocations inside the Constructor of the `struct` (and deletion inside the Destructor), otherwise, you can leak memory if somethings throws inside this (or any other) function before calling `free`.

Reply:
Unfortunately, I cannot move these inside the constructors as the allocations are dependent on the data from the buffer_size_ext cusparse function and hence can only be allocated after that function has been called.

Reply:
Then at least move the `free` to the destructor of the `struct`, that way, it should still be cleaned up, even if something throws inside this or another function.

Reply:
The destructor of the TRS solver should be called in case some function throws, right ? The `free` is in the call to that destructor, within the `clear` kernel. I checked an atleast here for this struct (maybe I should change the struct somehow) the destructor is not called even if I add one.

Reply:
That is correct, currently, it should be called in the destructor of the TRS class.
But usually, classes and structs should be responsible for the memory that is allocated inside them, which is why I would prefer to delete the memory in the destructor of the `struct`.
That also prevents programming errors (e.g. if someone wants to extend the current state, and forgets a single `clear` or `free`).

The destructor of the struct as it is now will only (if at all) be called when the whole program terminates since it is as global variable.